Sources/AWSLambdaDeploymentDescriptor/DeploymentDescriptorBuilder.swift

```diff
@@ -0,0 +1,424 @@
+// MARK: Function resource
+public struct Function {
+    let name: String
+    let architecture: Architectures
+    let eventSources: [Resource<EventSourceType>]
+    let environment: [String: String]
+    
+    private init(
+        name: String,
+        architecture: Architectures = Architectures.defaultArchitecture(),
+        eventSources: [Resource<EventSourceType>] = [],
+        environment: [String: String] = [:]
+    ) {
+        self.name = name
+        self.architecture = architecture
+        self.eventSources = eventSources
+        self.environment = environment
+    }
+    public init(
+        name: String,
+        architecture: Architectures = Architectures.defaultArchitecture(),
+        @FunctionBuilder _ builder: () -> (EventSources, [String: String])
+    ) {
+        
+        let (eventSources, environmentVariables) = builder()
+        let samEventSource: [Resource<EventSourceType>] = eventSources.samEventSources()
+        self.init(
+            name: name,
+            architecture: architecture,
+            eventSources: samEventSource,
+            environment: environmentVariables)
+    }
+    
+    // this method fails when the package does not exist at path
+    internal func resources() -> [Resource<ResourceType>] {
+        
+        let functionResource = [
+            Resource<ResourceType>.serverlessFunction(
+                name: self.name,
+                architecture: self.architecture,
+                codeUri: packagePath(),
+                eventSources: self.eventSources,
+                environment: SAMEnvironmentVariable(self.environment))
+        ]
+        let additionalQueueResources = collectQueueResources()
+        
+        return functionResource + additionalQueueResources
+    }
+    
+    // compute the path for the lambda archive
+    private func packagePath() -> String {
+        
+        // propose a default path unless the --archive-path argument was used
+        // --archive-path argument value must match the value given to the archive plugin --output-path argument
+        var lambdaPackage =
+        ".build/plugins/AWSLambdaPackager/outputs/AWSLambdaPackager/\(self.name)/\(self.name).zip"
+        if let optIdx = CommandLine.arguments.firstIndex(of: "--archive-path") {
+            if CommandLine.arguments.count >= optIdx + 1 {
+                let archiveArg = CommandLine.arguments[optIdx + 1]
+                lambdaPackage = "\(archiveArg)/\(self.name)/\(self.name).zip"
+            }
+        }
+        
+        // check the ZIP file exists
+        if !FileManager.default.fileExists(atPath: lambdaPackage) {
+            let msg = "Error: package does not exist at path: \(lambdaPackage)"
+            
+            // when running in a unit test, return a marker value
+            // otherwise, fail abruptly
+            if !Thread.current.isRunningXCTest {
+                fatalError(msg)
+            } else {
+                lambdaPackage = "ERROR"
+            }
```

Comment:
would throwing an exception be nicer here?

Reply:
done